File: sdk/src/governor.ts
import {
  classifyToolCall,
  sessionEventKind,
  DEFAULT_RULES,
  type Classification,
  type ClassifierRules,
  type Decision,
} from "./classifier.js";
import { MemorySessionStore, DEFAULT_WINDOWS, type SessionWindows } from "./session.js";

/** A tool call to govern. `args` may be a structured object or a raw JSON string. */
export interface ToolCall {
  name: string;
  args?: unknown;
}

/** What an approval handler receives when a call is HELD. */
export interface HoldRequest {
  name: string;
  args: unknown;
  classification: Classification;
  sessionKey: string;
}

/** Operator verdict for a held call. */
export type Verdict = "approve" | "deny";

export interface GateContext {
  /** Groups velocity counters (e.g. an agent id). Defaults to a single session. */
  sessionKey?: string;
  /**
   * Called when a call is HELD. Return "approve" to let it run, "deny" to block.
   * Omit it and held calls fall back to {@link PhinqConfig.defaultOnHold}.
   */
  onHold?: (req: HoldRequest) => Verdict | Promise<Verdict>;
  /** Auto-deny if the handler doesn't answer in time. Overrides the config default. */
  holdTimeoutMs?: number;
}

export type GateResolution = "allowed" | "approved" | "denied" | "timed_out";

export interface GateResult {
  /** The single thing the caller must honor: run the tool or not. */
  allowed: boolean;
  decision: Decision;
  classification: Classification;
  resolution: GateResolution;
}

export interface AuditEntry {
  ts: string;
  name: string;
  action_class: string;
  decision: Decision;
  triggers: string[];
  allowed: boolean;
  resolution: GateResolution;
  /** Byte length of the serialized arguments — never the arguments themselves. */
  args_bytes: number;
}

export interface PhinqConfig {
  /** Thresholds + tool→class overrides. Defaults to the standard rule set. */
  rules?: ClassifierRules;
  /** Rolling-window sizes for velocity triggers. */
  windows?: SessionWindows;
  /** What to do with a HELD call when no `onHold` handler is supplied. Default "deny". */
  defaultOnHold?: Verdict;
  /** Default auto-deny timeout for `onHold`. Default 240_000 ms (matches the proxy). */
  holdTimeoutMs?: number;
  /** Optional sink for one entry per governed call. Arguments are never included. */
  onAudit?: (entry: AuditEntry) => void;
}

/**
 * In-process governance for a TS agent. Wraps the same deterministic classifier
 * the Phinq proxy uses, adds rolling-window velocity tracking, and gates a tool
 * call on an operator verdict — all without a network hop. Unlike the proxy
 * (which intercepts the model's *proposed* call), `gate()` sits at the point of
 * *actual execution*, so a denial cannot be worked around by the agent.
 */
export class PhinqGovernor {
  private readonly rules: ClassifierRules;
  private readonly sessions: MemorySessionStore;
  private readonly defaultOnHold: Verdict;
  private readonly holdTimeoutMs: number;
  private readonly onAudit?: (entry: AuditEntry) => void;

  constructor(config: PhinqConfig = {}) {
    this.rules = config.rules ?? DEFAULT_RULES;
    this.sessions = new MemorySessionStore(config.windows ?? DEFAULT_WINDOWS);
    this.defaultOnHold = config.defaultOnHold ?? "deny";
    this.holdTimeoutMs = config.holdTimeoutMs ?? 240_000;
    this.onAudit = config.onAudit;
  }

  /** Classify without recording or gating — pure inspection of a single call. */
  classify(call: ToolCall, sessionKey = "default"): Classification {
    return classifyToolCall(
      { name: call.name, argumentsJson: serializeArgs(call.args) },
      this.sessions.counts(sessionKey),
      this.rules
    );
  }

  /**
   * Classify, record velocity, and resolve a verdict. Returns `allowed`: honor
   * it by running or skipping the tool. ALLOW resolves instantly; HOLD consults
   * `onHold` (or the configured default), auto-denying on timeout.
   */
  async gate(call: ToolCall, ctx: GateContext = {}): Promise<GateResult> {
    const sessionKey = ctx.sessionKey ?? "default";
    const counts = this.sessions.counts(sessionKey);
    const classification = classifyToolCall(
      { name: call.name, argumentsJson: serializeArgs(call.args) },
      counts,
      this.rules
    );

    // Record this call's velocity contribution for subsequent calls in the window.
    const kind = sessionEventKind(call.name);
    if (kind) this.sessions.record(sessionKey, kind);

    let allowed: boolean;
    let resolution: GateResolution;

    if (classification.decision === "ALLOW") {
      allowed = true;
      resolution = "allowed";
    } else if (!ctx.onHold) {
      allowed = this.defaultOnHold === "approve";
      resolution = allowed ? "approved" : "denied";
    } else {
      const verdict = await withTimeout(
        Promise.resolve(ctx.onHold({ name: call.name, args: call.args, classification, sessionKey })),
        ctx.holdTimeoutMs ?? this.holdTimeoutMs
      );
      allowed = verdict === "approve";
      resolution = verdict === "approve" ? "approved" : verdict === "deny" ? "denied" : "timed_out";
    }

    const result: GateResult = { allowed, decision: classification.decision, classification, resolution };
    this.onAudit?.({
      ts: new Date().toISOString(),
      name: call.name,
      action_class: classification.action_class,
      decision: classification.decision,
      triggers: classification.triggers,
      allowed,
      resolution,
      args_bytes: serializeArgs(call.args)?.length ?? 0,
    });
    return result;
  }
}

/** Stringify structured args for the arg-inspecting classifier; pass strings through. */
export function serializeArgs(args: unknown): string | undefined {
  if (args === undefined || args === null) return undefined;
  if (typeof args === "string") return args;
  try {
    return JSON.stringify(args);
  } catch {
    return undefined;
  }
}

/** Resolve the verdict, or "timeout" if it doesn't arrive in time. */
async function withTimeout(
  p: Promise<Verdict>,
  ms: number
): Promise<Verdict | "timeout"> {
  let timer: ReturnType<typeof setTimeout>;
  const timeout = new Promise<"timeout">((resolve) => {
    timer = setTimeout(() => resolve("timeout"), ms);
  });
  try {
    return await Promise.race([p, timeout]);
  } finally {
    clearTimeout(timer!);
  }
}

File: python/src/phinq/governor.py
"""In-process gate for tool execution — Python port of sdk/src/governor.ts.

    from phinq import PhinqGovernor

    governor = PhinqGovernor()
    result = governor.gate("delete_file", {"path": "/tmp/x"},
                           on_hold=lambda req: input(f"approve {req.name}? ") or "deny")
    if result.allowed:
        run_tool()

Unlike the proxy (which intercepts the model's *proposed* call), `gate()` sits
at the point of execution — the last line of defense before a side effect.
"""
from __future__ import annotations

import hashlib
import json
import time
from concurrent.futures import ThreadPoolExecutor, TimeoutError as FutureTimeout
from dataclasses import dataclass, field
from typing import Any, Callable

from .classifier import (
    Classification,
    ClassifierRules,
    DEFAULT_RULES,
    SessionCounts,
    classify_tool_call,
    session_event_kind,
)


@dataclass
class SessionWindows:
    """Rolling-window sizes (minutes) for velocity triggers."""

    window_minutes: int = 60
    error_window_minutes: int = 10


class MemorySessionStore:
    """In-memory rolling-window counters, keyed by session.

    `counts()` returns the window state BEFORE the current call, matching the
    classifier's contract (thresholds compare against prior activity).
    """

    def __init__(self, windows: SessionWindows | None = None) -> None:
        self.windows = windows or SessionWindows()
        self._events: dict[str, list[tuple[str, float]]] = {}

    def record(self, key: str, kind: str, now: float | None = None) -> None:
        now = time.time() if now is None else now
        self._events.setdefault(key, []).append((kind, now))
        self._prune(key, now)

    def counts(self, key: str, now: float | None = None) -> SessionCounts:
        now = time.time() if now is None else now
        self._prune(key, now)
        window_start = now - self.windows.window_minutes * 60
        error_start = now - self.windows.error_window_minutes * 60
        sends = deletes = 0
        recent_error = False
        for kind, ts in self._events.get(key, []):
            if kind == "send" and ts >= window_start:
                sends += 1
            elif kind == "delete" and ts >= window_start:
                deletes += 1
            elif kind == "error" and ts >= error_start:
                recent_error = True
        return SessionCounts(sends=sends, deletes=deletes, recent_error=recent_error)

    def _prune(self, key: str, now: float) -> None:
        longest = max(self.windows.window_minutes, self.windows.error_window_minutes)
        cutoff = now - longest * 60
        kept = [(k, t) for k, t in self._events.get(key, []) if t >= cutoff]
        if kept:
            self._events[key] = kept
        else:
            self._events.pop(key, None)


def session_key_from(identifier: str) -> str:
    """Hash an opaque session identifier — never store it raw."""
    return hashlib.sha256(identifier.encode()).hexdigest()


@dataclass
class HoldRequest:
    name: str
    args: Any
    classification: Classification
    session_key: str


@dataclass
class GateResult:
    allowed: bool
    decision: str
    classification: Classification
    resolution: str  # "allowed" | "approved" | "denied" | "timed_out"


@dataclass
class AuditEvent:
    ts: str
    name: str
    action_class: str
    decision: str
    triggers: list[str]
    allowed: bool
    resolution: str
    args_bytes: int


def _serialize_args(args: Any) -> str | None:
    if args is None:
        return None
    if isinstance(args, str):
        return args
    try:
        return json.dumps(args)
    except (TypeError, ValueError):
        return None


class PhinqGovernor:
    """Deterministic classifier + rolling-window velocity + human-in-the-loop gate."""

    def __init__(
        self,
        rules: ClassifierRules | None = None,
        windows: SessionWindows | None = None,
        default_on_hold: str = "deny",
        hold_timeout_s: float = 240.0,
        on_audit: Callable[[AuditEvent], None] | None = None,
    ) -> None:
        self.rules = rules or DEFAULT_RULES
        self.sessions = MemorySessionStore(windows)
        self.default_on_hold = default_on_hold
        self.hold_timeout_s = hold_timeout_s
        self.on_audit = on_audit

    def classify(self, name: str, args: Any = None, session_key: str = "default") -> Classification:
        return classify_tool_call(
            name, _serialize_args(args), self.sessions.counts(session_key), self.rules
        )

    def record_error(self, session_key: str = "default") -> None:
        """Mark a failure in this session — arms the AFTER_ERROR_BULK trigger."""
        self.sessions.record(session_key, "error")

    def gate(
        self,
        name: str,
        args: Any = None,
        *,
        session_key: str = "default",
        on_hold: Callable[[HoldRequest], str] | None = None,
        hold_timeout_s: float | None = None,
    ) -> GateResult:
        """Classify and gate one tool execution.

        `on_hold` receives a HoldRequest and returns "approve" or "deny".
        It runs with a timeout; no answer in time means deny ("timed_out").
        """
        args_json = _serialize_args(args)
        classification = classify_tool_call(
            name, args_json, self.sessions.counts(session_key), self.rules
        )

        kind = session_event_kind(name)
        if kind:
            self.sessions.record(session_key, kind)

        if classification.decision == "ALLOW":
            allowed, resolution = True, "allowed"
        elif on_hold is None:
            allowed = self.default_on_hold == "approve"
            resolution = "approved" if allowed else "denied"
        else:
            req = HoldRequest(name=name, args=args, classification=classification, session_key=session_key)
            timeout = self.hold_timeout_s if hold_timeout_s is None else hold_timeout_s
            with ThreadPoolExecutor(max_workers=1) as pool:
                future = pool.submit(on_hold, req)
                try:
                    verdict = future.result(timeout=timeout)
                except FutureTimeout:
                    verdict = None
                except Exception:
                    verdict = None
            if verdict == "approve":
                allowed, resolution = True, "approved"
            elif verdict == "deny":
                allowed, resolution = False, "denied"
            else:
                allowed, resolution = False, "timed_out"

        result = GateResult(
            allowed=allowed,
            decision=classification.decision,
            classification=classification,
            resolution=resolution,
        )
        if self.on_audit:
            self.on_audit(
                AuditEvent(
                    ts=time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                    name=name,
                    action_class=classification.action_class.value,
                    decision=classification.decision,
                    triggers=list(classification.triggers),
                    allowed=allowed,
                    resolution=resolution,
                    args_bytes=len(args_json) if args_json else 0,
                )
            )
        return result
